Mathlib/Algebra/Star/Positive.lean

```diff
@@ -0,0 +1,178 @@
+/-
+Copyright (c) 2023 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import Mathlib.Algebra.Star.Order
+import Mathlib.Algebra.Star.SelfAdjoint
+import Mathlib.GroupTheory.Submonoid.Operations
+
+/-!
+# Positive elements in a star ordered ring
+
+This file defines the `AddSubmonoid` of positive elements `StarOrderedRing.positive R` in a
+star ordered ring `R` as generated by the elements of the form `star s * s`. It's carrier is
+propositionally equal to `{ x : R | 0 ≤ x }`. Even though this technically corresponds to the
+nonnegative elements, the "positive" terminology is standard throughout the literature (at least
+in C⋆-algebra theory and operator theory), so we choose to prefer that.
+
+The advantage of using this over other definitions of positivity is that it allows us to unify
+several different concepts under a single umbrella. For example, one might be tempted to consider
+the collection of bounded linear operators on an Hilbert space and define positivity as
+`∀ x, 0 ≤ ⟪T x, x⟫`, but this doesn't generalize nicely to C⋆-algebras. Alternatively, one could
+define positivity as those self-adjoint elements with nonnegative spectrum, but this requires the
+algebra to be unital. Our definition herein conveniently sidesteps these issues.
+
+We place everything in the `StarOrderedRing` namespace in order to avoid collisions or confusion
+with other parts of the library.
+
+## Main definitions
+
+* `StarOrderedRing.positive R`: The additive submonoid of a star ordered ring `R` generated by
+  elements of the form `star s * s` for `s : R`.
+* `StarOrderedRing.positive.selfAdjoint : positive R →+ selfAdjoint R`: the identity map as an
+  `AddMonoidHom`.
+* `StarOrderedRing.positive.instAddMonoidWithOne`: Makes `positive R` into an `AddMonoidWithOne`
+* `StarOrderedRing.positive.instCanonicallyOrderedAddMonoid`: Makes `positive R` into a
+  `CanonicallyOrderedAddMonoid` which avoids duplication of some lemmas (e.g., `zero_le`)
+
+-/
+
+universe u
+
+variable {R : Type u}
+
+namespace StarOrderedRing
+
+def positive (R : Type u) [NonUnitalSemiring R] [PartialOrder R] [StarOrderedRing R] :
+    AddSubmonoid R :=
+  AddSubmonoid.closure (Set.range fun s : R => star s * s)
```

Comment:
Just in case anyone sees this PR but not the Zulip thread; I'd argue we want instead
```
def AddSubmonoid.nonneg (M : Type u) [OrderedAddCommMonoid M] : AddSubmonoid M where
  carrier := {m : M | 0 ≤ m}
  zero_mem := le_rfl
  add_mem := add_nonneg
```
which as I understand it is strictly more general. This could probably go in `Mathlib/Algebra/Order/Nonneg/Ring.lean`.